runtime/pkg/duckdbreplicator/db.go

```diff
@@ -0,0 +1,1202 @@
+
+	BackupProvider *BackupProvider
+
+	// ReadSettings are settings applied the read duckDB handle.
+	ReadSettings map[string]string
+	// WriteSettings are settings applied the write duckDB handle.
+	WriteSettings map[string]string
+	// InitQueries are the queries to run when the database is first created.
+	InitQueries []string
+
+	Logger *slog.Logger
+}
+
+// TODO :: revisit this logic
+func (d *DBOptions) ValidateSettings() error {
+	read := &settings{}
+	err := mapstructure.Decode(d.ReadSettings, read)
+	if err != nil {
+		return fmt.Errorf("read settings: %w", err)
+	}
+
+	write := &settings{}
+	err = mapstructure.Decode(d.WriteSettings, write)
+	if err != nil {
+		return fmt.Errorf("write settings: %w", err)
+	}
+
+	// no memory limits defined
+	// divide memory equally between read and write
+	if read.MaxMemory == "" && write.MaxMemory == "" {
+		connector, err := duckdb.NewConnector("", nil)
+		if err != nil {
+			return fmt.Errorf("unable to create duckdb connector: %w", err)
+		}
+		defer connector.Close()
+		db := sql.OpenDB(connector)
+		defer db.Close()
+
+		row := db.QueryRow("SELECT value FROM duckdb_settings() WHERE name = 'max_memory'")
+		var maxMemory string
+		err = row.Scan(&maxMemory)
+		if err != nil {
+			return fmt.Errorf("unable to get max_memory: %w", err)
+		}
+
+		bytes, err := humanReadableSizeToBytes(maxMemory)
+		if err != nil {
+			return fmt.Errorf("unable to parse max_memory: %w", err)
+		}
+
+		read.MaxMemory = fmt.Sprintf("%d bytes", int64(bytes)/2)
+		write.MaxMemory = fmt.Sprintf("%d bytes", int64(bytes)/2)
+	}
+
+	if read.MaxMemory == "" != (write.MaxMemory == "") {
+		// only one is defined
+		var mem string
+		if read.MaxMemory != "" {
+			mem = read.MaxMemory
+		} else {
+			mem = write.MaxMemory
+		}
+
+		bytes, err := humanReadableSizeToBytes(mem)
+		if err != nil {
+			return fmt.Errorf("unable to parse max_memory: %w", err)
+		}
+
+		read.MaxMemory = fmt.Sprintf("%d bytes", int64(bytes)/2)
+		write.MaxMemory = fmt.Sprintf("%d bytes", int64(bytes)/2)
+	}
+
+	var readThread, writeThread int
+	if read.Threads != "" {
+		readThread, err = strconv.Atoi(read.Threads)
+		if err != nil {
+			return fmt.Errorf("unable to parse read threads: %w", err)
+		}
+	}
+	if write.Threads != "" {
+		writeThread, err = strconv.Atoi(write.Threads)
+		if err != nil {
+			return fmt.Errorf("unable to parse write threads: %w", err)
+		}
+	}
+
+	if readThread == 0 && writeThread == 0 {
+		connector, err := duckdb.NewConnector("", nil)
+		if err != nil {
+			return fmt.Errorf("unable to create duckdb connector: %w", err)
+		}
+		defer connector.Close()
+		db := sql.OpenDB(connector)
+		defer db.Close()
+
+		row := db.QueryRow("SELECT value FROM duckdb_settings() WHERE name = 'threads'")
+		var threads int
+		err = row.Scan(&threads)
+		if err != nil {
+			return fmt.Errorf("unable to get threads: %w", err)
+		}
+
+		read.Threads = strconv.Itoa((threads + 1) / 2)
+		write.Threads = strconv.Itoa(threads / 2)
+	}
```

Comment:
Actually in past we have seen(also discussed this with duckDB team) that duckDB works best with a particular memory to threads ratio. Not dividing threads will mean we reduce our current ratio (which I think we ended up at after some permutations) to half. Though difficult to say without measuring things but I suspect it can be problematic ?

Reply:
Okay if you are worried about that, then we can keep dividing the threads. Beware that `int(threads/2)` might lead to 0 threads being assigned if `threads` is `1`.